oak-blob-plugins/src/main/java/org/apache/jackrabbit/oak/plugins/blob/datastore/AbstractDataStoreService.java

```diff
@@ -60,70 +67,103 @@
     public static final String PROP_CACHE_SIZE = "cacheSizeInMB";
     private static final String DESCRIPTION = "oak.blobstore.description";
 
-    private ServiceRegistration reg;
-
-    private Registration mbeanReg;
-
-    private Logger log = LoggerFactory.getLogger(getClass());
+    private static final Logger log = LoggerFactory.getLogger(AbstractDataStoreService.class);
 
     @Reference
     private StatisticsProvider statisticsProvider;
 
-    private DataStoreBlobStore dataStore;
+    protected Closer closer = Closer.create();
 
     protected void activate(ComponentContext context, Map<String, Object> config) throws RepositoryException {
         // change to mutable map. may be modified in createDS call
         config = Maps.newHashMap(config);
+
         DataStore ds = createDataStore(context, config);
+        registerDataStore(context, config, ds, getStatisticsProvider(), getDescription(), closer);
+    }
+
+    protected void deactivate() throws DataStoreException {
+        if (null != closer) {
+            synchronized (this) {
+                if (null != closer) {
+                    closeQuietly(closer);
+                    closer = null;
+                }
+            }
+        }
+    }
+
+    protected abstract DataStore createDataStore(ComponentContext context, Map<String, Object> config);
+
+    protected StatisticsProvider getStatisticsProvider(){
+        return statisticsProvider;
+    }
+
+    public static DataStore registerDataStore(
+            @Nonnull ComponentContext context,
+            @Nonnull Map<String, Object> config,
+            @Nullable DataStore ds,
+            @Nonnull StatisticsProvider statisticsProvider,
+            @Nonnull String[] description,
+            @Nonnull Closer closer
+    ) throws RepositoryException {
+        if (null == ds) {
+            // Deferred registration - child class should call registerDataStore later
+            return ds;
+        }
+
+        Closeables closeables = new Closeables(closer);
+
         boolean encodeLengthInId = PropertiesUtil.toBoolean(config.get(PROP_ENCODE_LENGTH), true);
         int cacheSizeInMB = PropertiesUtil.toInteger(config.get(PROP_CACHE_SIZE), DataStoreBlobStore.DEFAULT_CACHE_SIZE);
 
         String homeDir = lookup(context, PROP_HOME);
         if (config.containsKey(PATH) && !Strings.isNullOrEmpty((String) config.get(PATH))) {
             log.info("Initializing the DataStore with path [{}]", config.get(PATH));
+            homeDir = (String) config.get(PATH);
```

Comment:
This is wrong and completely changes the location of the datastore folder. This is a byproduct of not properly calling _init()_ on the delegate DataStores

Reply:
Yes, that it does.  I assumed it was supposed to be changed based on the log message above.  Why are we even checking for the PATH configuration value, and why are we logging that we are initializing the data store with that value, if we actually are not doing that?  If we don't change the value here we are basically logging a misleading message.

Reply:
Well it was not being changed earlier because it would change the location and the logging was added to precisely remove the confusion of where  exactly the path is initialized. Do take a look at the FileDataStore.init(). 
We cannot change it because it would cause existing systems to not find the DataStore.